### ISSUE TITLE
invalid certificate file doesnt raise error in 112.24.03

### ISSUE THREAD

**Issue by avsm** [janestreet/async_ssl#6]
In older Async_ssl, it would raise an error if the certificate file was invalid:

```
((pid 79800) (thread_id 0) (2015-04-17 15:24:59.904484+01:00)
 "unhandled exception in Async scheduler"
 ("unhandled exception"
  ((lib/monitor.ml.Error_
    ((exn
      (Failure
       "Could not load certificate file @ /Users/avsm/src/git/avsm/ocaml-cohttp/server.pem"))
     (backtrace
      ("Raised at file \"pervasives.ml\", line 30, characters 22-33"
       "Called from file \"lib/deferred.ml\", line 15, characters 64-67"
       "Called from file \"lib/jobs.ml\", line 214, characters 10-13" ""))
     (monitor
      (((name Tcp.collect_errors) (here ()) (id 17) (has_seen_error true)
        (is_detached true) (kill_index 0))))))
   (Pid 79800))))
```

Unfortunately with Async_ssl 112.24.03, an incorrect file just hangs the server process instead...

**Comment by trefis** [comment]
ping @tari3x

**Comment by avsm** [comment]
There's a simple testcase in mirage/ocaml-conduit#59 for use with Conduit_async

**Comment by tari3x** [comment]
Will look at it first thing Tuesday.

**Comment by tari3x** [comment]
Certificate check failures are now included in the Or_error that is the result of [Ssl.client] and [Ssl.server], but I see that Conduit_async_ssl ignores those values.  I would also suggest waiting on [Connection.closed] since this will notify you of more errors that might happen after the handshake is complete.

I changed the semantics of these functions without changing their types enough since I was assuming that async_ssl is only used in the jane tree - apologies for breaking your world!

The reason you hang is that a certificate load error does not close the pipes. I'm reluctant to change this since this doesn't really reduce the confusion (now the server would silently terminate instead of hanging). 

As an aside, async_ssl changed a lot recently - I'd suggest looking at the comments regarding certificate checking in the mli. In short, an important part of certificate validation must be done by the caller of the library - if you don't do it, your achieved level of security is very weak.

For clarity, in the previous version of async_ssl, the certificates were not checked at all - this is the default OpenSSL setting that everybody overlooks.

**Comment by hhugo** [comment]
What's the status of this ?

**Comment by kkazuo** [comment]
I tried and here is the results.

Version:
```
% opam list | grep async
async                   v0.13.0     Monadic concurrency library
async_kernel            v0.13.0     Monadic concurrency library
async_rpc_kernel        v0.13.0     Platform-independent core of Async RPC library
async_ssl               v0.13.0     An Async-pipe-based interface with OpenSSL
async_unix              v0.13.0     Monadic concurrency library
cohttp-async            2.5.1       CoHTTP implementation for the Async concurrency library
conduit-async           2.1.0       A network connection establishment library for Async
```
Results:
```
% cohttp-curl-async https://expired.badssl.com
(monitor.ml.Error
 (Ssl_error
  ("error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed")
  src/ssl.ml:237:4)
 ("Raised at file \"src/error.ml\" (inlined), line 8, characters 14-30"
  "Called from file \"src/error.ml\" (inlined), line 5, characters 2-50"
  "Called from file \"src/ssl.ml\", line 237, characters 4-74"
  "Called from file \"src/deferred0.ml\", line 54, characters 64-69"
  "Called from file \"src/job_queue.ml\", line 167, characters 6-47"
  "Caught by monitor ssl_pipe"))
```
```
% cohttp-curl-async https://wrong.host.badssl.com
cohttp-curl-async: [INFO] accept-ranges: bytes
cohttp-curl-async: [INFO] cache-control: no-store
cohttp-curl-async: [INFO] connection: close
cohttp-curl-async: [INFO] content-length: 500
cohttp-curl-async: [INFO] content-type: text/html
cohttp-curl-async: [INFO] date: Thu, 16 Apr 2020 15:06:08 GMT
cohttp-curl-async: [INFO] etag: "5e79513a-1f4"
cohttp-curl-async: [INFO] last-modified: Tue, 24 Mar 2020 00:15:54 GMT
cohttp-curl-async: [INFO] server: nginx/1.10.3 (Ubuntu)
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="shortcut icon" href="/icons/favicon-red.ico"/>
  <link rel="apple-touch-icon" href="/icons/icon-red.png"/>
  <title>wrong.host.badssl.com</title>
  <link rel="stylesheet" href="/style.css">
  <style>body { background: red; }</style>
</head>
<body>
<div id="content">
  <h1 style="font-size: 12vw;">
    wrong.host.<br>badssl.com
  </h1>
</div>

</body>
</html>
```
```
% cohttp-curl-async https://self-signed.badssl.com
(monitor.ml.Error
 (Ssl_error
  ("error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed")
  src/ssl.ml:237:4)
 ("Raised at file \"src/error.ml\" (inlined), line 8, characters 14-30"
  "Called from file \"src/error.ml\" (inlined), line 5, characters 2-50"
  "Called from file \"src/ssl.ml\", line 237, characters 4-74"
  "Called from file \"src/deferred0.ml\", line 54, characters 64-69"
  "Called from file \"src/job_queue.ml\", line 167, characters 6-47"
  "Caught by monitor ssl_pipe"))
```
```
% cohttp-curl-async https://untrusted-root.badssl.com
(monitor.ml.Error
 (Ssl_error
  ("error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed")
  src/ssl.ml:237:4)
 ("Raised at file \"src/error.ml\" (inlined), line 8, characters 14-30"
  "Called from file \"src/error.ml\" (inlined), line 5, characters 2-50"
  "Called from file \"src/ssl.ml\", line 237, characters 4-74"
  "Called from file \"src/deferred0.ml\", line 54, characters 64-69"
  "Called from file \"src/job_queue.ml\", line 167, characters 6-47"
  "Caught by monitor ssl_pipe"))
```
```
% cohttp-curl-async https://revoked.badssl.com
cohttp-curl-async: [INFO] accept-ranges: bytes
cohttp-curl-async: [INFO] cache-control: no-store
cohttp-curl-async: [INFO] connection: close
cohttp-curl-async: [INFO] content-length: 575
cohttp-curl-async: [INFO] content-type: text/html
cohttp-curl-async: [INFO] date: Thu, 16 Apr 2020 15:06:59 GMT
cohttp-curl-async: [INFO] etag: "5e79513a-23f"
cohttp-curl-async: [INFO] last-modified: Tue, 24 Mar 2020 00:15:54 GMT
cohttp-curl-async: [INFO] server: nginx/1.10.3 (Ubuntu)
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="shortcut icon" href="/icons/favicon-red.ico"/>
  <link rel="apple-touch-icon" href="/icons/icon-red.png"/>
  <title>revoked.badssl.com</title>
  <link rel="stylesheet" href="/style.css">
  <style>body { background: red; }</style>
</head>
<body>
<div id="content">
  <h1 style="font-size: 10vw;">
    revoked.<br>badssl.com
  </h1>
</div>

<div id="footer">
  The leaf certificate for this site has been revoked.
</div>

</body>
</html>
```
```
% cohttp-curl-async https://pinning-test.badssl.com
cohttp-curl-async: [INFO] accept-ranges: bytes
cohttp-curl-async: [INFO] cache-control: no-store
cohttp-curl-async: [INFO] connection: close
cohttp-curl-async: [INFO] content-length: 573
cohttp-curl-async: [INFO] content-type: text/html
cohttp-curl-async: [INFO] date: Thu, 16 Apr 2020 15:07:27 GMT
cohttp-curl-async: [INFO] etag: "5e79513a-23d"
cohttp-curl-async: [INFO] last-modified: Tue, 24 Mar 2020 00:15:54 GMT
cohttp-curl-async: [INFO] server: nginx/1.10.3 (Ubuntu)
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="shortcut icon" href="/icons/favicon-red.ico"/>
  <link rel="apple-touch-icon" href="/icons/icon-red.png"/>
  <title>pinning-test.badssl.com</title>
  <link rel="stylesheet" href="/style.css">
  <style>body { background: red; }</style>
</head>
<body>
<div id="content">
  <h1>
    pinning-test.<br>badssl.com
  </h1>
</div>

<div id="footer">
  This site is preloaded with a bad HPKP pin starting in Chrome 48.
</div>

</body>
</html>
```

**Comment by tiash** [comment]
There are two independent issues here.

@avsm We're not able to reproduce the original issue of the server locking up if provided with an invalid server certificate. Please let us know if this is still an issue and how to reproduce.


@kkazuo This is a limitation of the underlying OpenSSL library. Host name validation is a recently added feature and not available in the versions of OpenSSL we need to support. See https://wiki.openssl.org/index.php/Hostname_validation
We do expose [Connection.peer_certificate] to access the remote server's certificate so you could do host name validation on this.
